File: sdk/python/src/alpine_sdk/transport/udp.py
"""UDP transport helper used by the Alpine SDK clients."""

import socket
from typing import Optional, Tuple


class UdpTransport:
    """Lightweight UDP socket wrapper."""

    def __init__(self, remote: Tuple[str, int], local_host: str, local_port: int) -> None:
        self._remote = remote
        self._socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._socket.bind((local_host, local_port))

    def send(self, payload: bytes, destination: Optional[Tuple[str, int]] = None) -> None:
        """Sends raw bytes to the provided destination or the configured peer."""

        target = destination or self._remote
        self._socket.sendto(payload, target)

    def close(self) -> None:
        """Closes the underlying socket."""

        self._socket.close()

File: sdk/ts/src/transport/udp.ts
import * as dgram from "dgram";

import { AlpineSdkError } from "../errors";

export interface UdpTransportOptions {
  remoteHost: string;
  remotePort: number;
  localPort?: number;
}

/**
 * Simple UDP helper used by the SDK clients.
 */
export class UdpTransport {
  private readonly socket: dgram.Socket;

  constructor(private readonly options: UdpTransportOptions) {
    this.socket = dgram.createSocket("udp4");
    this.socket.bind(options.localPort ?? 0);
  }

  /**
   * Sends the provided payload to the configured remote backend.
   */
  send(payload: Buffer): Promise<void> {
    return new Promise((resolve, reject) => {
      this.socket.send(
        payload,
        this.options.remotePort,
        this.options.remoteHost,
        (err: Error | null) => {
          if (err) {
            reject(new AlpineSdkError(`udp send failure: ${err.message}`));
          } else {
            resolve();
          }
        },
      );
    });
  }

  /**
   * Releases the socket resources.
   */
  close(): void {
    this.socket.close();
  }
}
